### DISCUSSION TITLE
Incorrect land / sea polygons

### DISCUSSION THREAD

**Discussion by dnesbitt61** [mapsforge/mapsforge#1485] [General]
Using downloaded maps from: https://download.mapsforge.org/maps/v5/north-america/us/ I am seeing incorrect land/sea when I select 2 map files (in this case Maryland and Delaware) using the latest Cruiser app.

![image](https://github.com/mapsforge/vtm/assets/1838768/9e4a6f6b-7f3a-4a37-b445-810e70c020d9)

If I use a simple style with only issea and nosea I get:

![image](https://github.com/mapsforge/vtm/assets/1838768/2a548418-7514-47a7-98c9-903380169109)

Style contents:

```
<?xml version="1.0" encoding="UTF-8"?>
<rendertheme xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" map-background="#dcd5d0"
    version="1" xmlns="http://mapsforge.org/renderTheme"
    xsi:schemaLocation="http://mapsforge.org/renderTheme https://raw.githubusercontent.com/mapsforge/vtm/master/resources/rendertheme.xsd">

    <!-- AREA STYLES -->
    <style-area fade="-1" fill="#6e91cd" id="water"/>

    <!-- STYLE ASSIGNMENT RULES -->

    <!-- Land vs. sea -->
    <m e="way" k="natural" v="issea|sea">
        <area mesh="true" use="water"/>
    </m>
    <m e="way" k="natural" v="nosea">
        <area mesh="true" fill="#dcd5d0"/>
    </m>

</rendertheme>

```

**Comment by devemux86** [comment]
Mapsforge has similar rendering (see images), so this does not seem a VTM case.

Perhaps the creation of these maps caused some missing polygons? It also happens with 1 map.
We could try different map-writer options, for example `polygon-clipping=false`.

Map creation uses data and bounds from Geofabrik:
https://download.geofabrik.de/north-america/us/maryland.html
https://download.geofabrik.de/north-america/us/delaware.html

![Cruiser_1](https://github.com/mapsforge/vtm/assets/3484020/dab6cd44-e099-42d9-919a-da453d41bb76)
![Cruiser_2](https://github.com/mapsforge/vtm/assets/3484020/a6a5acd2-cdeb-488d-ab1c-7749f0ac18bb)

**Comment by devemux86** [comment]
@dnesbitt61 issue moved to Mapsforge repository as it appears to be related to maps.

It happens with even 1 map (see image above).

**Comment by dnesbitt61** [comment]
I tried with polygon-clipping=false and still see the issue. Also, the map file sizes are much larger (Maryland went from 129MB to over 1GB). I have seen issues like this caused by using shape2osm.py in map-creator. I have used  ogr2osm.py instead in the past to get around issues like this. 

The reason I think this issue is due to VTM drawing code is that data created over a year ago has triangulation issues with the latest Cruiser (VTM) and on Android (Mapsforge-VTM version 0.18, 0.19 or 0.20) when multiple .map files exist. The same data does not have these triangulation issues when using version 0.17. I use mesh=true for all area features in the style file.

**Comment by devemux86** [comment]
@dnesbitt61 thanks for the tests.

The issue also appears with Mapsforge map library and with just 1 map.
So it does not seem related to VTM or overlapping maps.
Or do you mean something else?

Map creation process used in [mapsforge-creator](https://github.com/mapsforge/mapsforge-creator) is explained with details in the [map creation guide](https://github.com/mapsforge/mapsforge/blob/master/docs/MapCreation.md).

shape2osm converts the land shp, after ogr2ogr cuts the land with bounds from [poly2bb](https://github.com/mapsforge/mapsforge-creator/blob/master/poly2bb.pl),
using the Geofabrik polygons with an extra buffer. Perhaps this buffer should be larger?

If the ogr2ogr process and its buffer is not the problem, then we could see the shape2osm.
However shape2osm does more than just a conversion, also adding needed land osm tags.
Does the other tool can be used or adapted to do the same things?

**Comment by dnesbitt61** [comment]
I will add a different issue to VTM repo showing the regression from 0.17 going to 0.18 with triangulation (when drawing data created over a year ago).

For map-creator, several years ago when I first started creating data for Mapsforge, I found myself always having to change the buffer (by trial and error often) to get rid of the land/sea rectangular cutouts. Some experimentation led me to switch to using ogr2osm.py. The github page for this is here: https://github.com/pnorman/ogr2osm. The program comments say: "By default tags will be naively copied from the input data. Hooks are provided so that, with a little python programming, you can translate the tags however you like."

**Comment by devemux86** [comment]
The buffer is used in ogr2ogr process but before the conversion to osm with another tool.
Have you tried a larger buffer with this map, does it improve anything?

> land/sea rectangular cutouts

Sea should not have cutouts as it is just a single rectangle below the map.

Land polygons from [OpenStreetMap Data](https://osmdata.openstreetmap.de/) have split and _not_ split version.
Does the _not_ split version works better?

Or you do not mean cutouts inside the land but at the bounds by the ogr2ogr buffer?
